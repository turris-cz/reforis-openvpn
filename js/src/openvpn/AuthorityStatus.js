/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React, { useEffect } from "react";
import PropTypes from "prop-types";

import { useAPIGet, Spinner } from "foris";

import API_URLs from "API";
import AuthorityMissing from "./AuthorityMissing";
import AuthorityReady from "./AuthorityReady";
import AuthorityBeingGenerated from "./AuthorityBeingGenerated";

AuthorityStatus.propTypes = {
    ws: PropTypes.object.isRequired,
};

export default function AuthorityStatus({ ws }) {
    const [authority, getAuthority] = useAPIGet(API_URLs.authority);
    useEffect(() => {
        getAuthority();
    }, [getAuthority]);

    if (authority.isLoading || !authority.data) {
        return <Spinner className="my-3 text-center" />;
    }

    if (authority.data.status === "missing") {
        return <AuthorityMissing onSuccess={getAuthority} />;
    }
    if (authority.data.status === "ready") {
        return <AuthorityReady onSuccess={getAuthority} />;
    }
    if (authority.data.status === "generating") {
        return <AuthorityBeingGenerated ws={ws} onSuccess={getAuthority} />;
    }
}
