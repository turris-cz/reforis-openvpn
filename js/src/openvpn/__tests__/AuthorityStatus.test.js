/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";
import { render, waitForElement, getByText } from "customTestRender";
import mockAxios from 'jest-mock-axios';

import { WebSockets } from "foris";

import AuthorityStatus from "../AuthorityStatus";

describe("<AuthorityStatus />", () => {
    let webSockets,
        componentContainer;

    beforeEach(() => {
        webSockets = new WebSockets();
        const { container } = render(<AuthorityStatus ws={webSockets} />);
        componentContainer = container;
    });

    it("should display spinner on loading", () => {
        expect(componentContainer).toMatchSnapshot();
    });

    it("should handle missing CA", async () => {
        mockAxios.mockResponse({data: {status: "missing"}});
        await waitForElement(() => getByText(componentContainer, "No certification authority"));
        expect(componentContainer).toMatchSnapshot();
    });

    it("should handle ready CA", async () => {
        mockAxios.mockResponse({data: {status: "ready"}});
        await waitForElement(() => getByText(componentContainer, "Certificate authority"));
        expect(componentContainer).toMatchSnapshot();
    });

    it("should handle generating CA", async () => {
        mockAxios.mockResponse({data: {status: "generating"}});
        await waitForElement(() => getByText(componentContainer, "Generating certification authority"));
        expect(componentContainer).toMatchSnapshot();
    });
});
