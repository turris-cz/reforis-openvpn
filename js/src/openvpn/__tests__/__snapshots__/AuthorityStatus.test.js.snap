// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AuthorityStatus /> should display spinner on loading 1`] = `
<div>
  <div
    class="spinner-wrapper my-3 text-center"
  >
    <div
      class="spinner-border "
      role="status"
    >
      <span
        class="sr-only"
      />
    </div>
    <div
      class="spinner-text"
    />
  </div>
</div>
`;

exports[`<AuthorityStatus /> should handle generating CA 1`] = `
<div>
  <h3>
    Generating certification authority
  </h3>
  <div
    class="spinner-wrapper my-3 text-center"
  >
    <div
      class="spinner-border "
      role="status"
    >
      <span
        class="sr-only"
      />
    </div>
    <div
      class="spinner-text"
    />
  </div>
  <p>
    Your certification authority is now being generated. It usually takes a few minutes. Settings will appear here automatically when the authority is ready.
  </p>
</div>
`;

exports[`<AuthorityStatus /> should handle missing CA 1`] = `
<div>
  <h3>
    No certification authority
  </h3>
  <p>
    Currently there is no OpenVPN certificate authority (CA). A CA is required to generate client certificates to authenticate to the OpenVPN server. To proceed you need to generate it first.
  </p>
  <button
    class="btn btn-primary "
    type="button"
  >
     
     
    Generate CA
  </button>
</div>
`;

exports[`<AuthorityStatus /> should handle ready CA 1`] = `
<div>
  <h3>
    Certificate authority
  </h3>
  <p>
    Your certificate authority (CA) is set up properly.
  </p>
  <button
    class="btn btn-primary "
    type="button"
  >
     
     
    Delete CA
  </button>
</div>
`;
